behavior/plans/diff.py

```diff
@@ -1,466 +1,288 @@
-        query_invocations = unified.loc[query_id]
-        if isinstance(query_invocations, pd.Series):
-            continue
-
-        node_ids: pd.Series = query_invocations["plan_node_id"]
-        node_counts: pd.Series = node_ids.value_counts()
-        assert (
-            node_counts.min() == node_counts.max()
-        ), f"Invalid node_id set for query_id: {query_id}.  Node_counts: {node_counts}"
-
-        assert (
-            query_invocations["rid"].value_counts().max() == 1
-        ), f"Found duplicate rids in query_invocations: {query_invocations}"
-
-        query_invocation_ids = set(pd.unique(query_invocations["invocation_id"]))
-        logger.info("Query ID: %s, Num invocations: %s", query_id, len(query_invocation_ids))
-        indexed_invocations = query_invocations.set_index("invocation_id", drop=False, inplace=False)
-
-        for invocation_id in query_invocation_ids:
-            invocation = indexed_invocations.loc[invocation_id]
-            if isinstance(invocation, pd.Series):
-                continue
+    Parameters
+    ----------
+    unified : Dataframe
+        Dataframe contains all the data that needs to be diferenced. The dataframe must folow
+        the DIFFERENCING_SCHEMA layout.
 
-            for rid, diffed_costs in diff_one_invocation(invocation).items():
-                records.append([rid] + diffed_costs.tolist())
-
-    diffed_cols = DataFrame(data=records, columns=["rid"] + DIFF_COLS)
-    diffed_cols.to_csv(logdir / "diffed_cols.csv", index=False)
-    diffed_cols.set_index("rid", drop=True, inplace=True)
-
-    return diffed_cols
+    diffed_matrixes : list[np.pdarray]
+        Output list to store diffed numpy matrixes.
+    """
+    # Here we assume that <query_id, statement_timestamp, pid> identifies a unique query session template.
+    # Grouping by <query_id, statement_timestamp, pid> will produce subframes of OUs for a given query session template.
+    invocation_groups = unified.groupby(by=["query_id", "statement_timestamp", "pid"], sort=False)
 
+    # We use apply() because we want to process the entire subframe as a single unit. transform() will not work
+    # for this because transform() may separate the columns.
+    invocation_groups.progress_apply(process_query_invocation, diffed_matrixes=diffed_matrixes)
 
-def save_results(diff_data_dir, ou_to_df, diffed_cols):
-    """Write the differenced data back to the original
-    DataFrame and save the new DataFrames to disk.
+
+def load_tscout_data(tscout_data_dir):
+    """
+    Load TScout data into dataframes.
 
     Parameters
     ----------
-    diff_data_dir : Path
-        Directory to save the differenced data.
-    ou_to_df : dict[str, DataFrame]
-        Map from operating-unit name to the original DataFrame.
-    diffed_cols : DataFrame
-        DataFrame of all differenced records, indexed by RID.
+    tscout_data_dir : Path
+        Data directory containing all the TScout raw data.
+
+    Returns:
```

Comment:
```suggestion
    Returns
```

CLion: Function 'load_tscout_data' does not have a parameter 'Returns'